src/library/library.cpp

```diff
@@ -692,6 +693,46 @@ bool Library::requestRelocateDir(const QString& oldDir, const QString& newDir) {
     return false;
 }
 
+#ifdef Q_OS_IOS
+void Library::requestRelocateiOSSandboxDirs() {
+    // TODO: If the user selects a different app sandbox than ours (which is
+    // possible via the file picker) relinking will point those directories to
+    // our sandbox. This is not a supported scenario, however, and we should probably
+    // guard against picking directories outside the Mixxx sandbox when adding
+    // new directories.
+
+    // TODO: Do we need to handle external track collections?
+
+    bool needsRescan = false;
+    QStringList rootDirs = m_pTrackCollectionManager->internalCollection()->getRootDirStrings();
+
+    for (const QString& dir : rootDirs) {
+        QString newDir = mixxx::updateiOSSandboxPath(dir);
+
+        if (dir == newDir) {
+            // Sandbox directory did not move
+            continue;
+        }
+
+        qInfo() << "Relinking music directory since iOS sandbox moved:"
+                << dir << "->" << newDir;
+
+        auto result = m_pTrackCollectionManager->relocateDirectory(dir, newDir);
+        if (result != DirectoryDAO::RelocateResult::Ok) {
+            qWarning() << "Could not relink music directory after iOS sandbox moved";
+            continue;
+        }
+
+        needsRescan = true;
```

Comment:
Did you verify a rescan is indeed required after relinking?
Various tests I ran showed it was not.
See also https://github.com/mixxxdj/mixxx/issues/12858#issuecomment-1961193790

Reply:
I think it was indeed required to make it work, if I remember right, but I'll check just to make sure I didn't miss anything (Though rescanning is probably still a good idea, the sandbox doesn't move that often, usually only when the app is updated, so I think we can afford to code a bit defensively here)

Reply:
Just tested it on a device and can confirm that a rescan is indeed needed, without it dragging a track from the library to the decks won't do anything.

Reply:
Mm do you know why? Maybe there is just a refresh of the library page required?

Reply:
What do you mean by that? This happens after reinstalling Mixxx, so the whole app is already relaunched.

Reply:
I think what @daschuer suspects is that the track view still shows old data.
Not sure if/how thats possible, but you may verify by switching to another library feature and back to **Tracks** for example to trigger a select(), then try again to load a track.

Of course, if load is still not possible after a second relaunch, the rescan is indeed required (still don't understand why..)

Reply:
Switching the view unfortunately doesn't change this. I did a quick database dump before and after rescanning (after reinstall + relink) and it looks like the relink doesn't update the track locations:

<img width="1571" alt="image" src="https://github.com/user-attachments/assets/6f353bdc-8ef3-4214-98d2-1d86fe8469f7">

The only entry in the pre-rescan/post-relink db that's updated to the new prefix is the `directories` entry:

```sql
INSERT INTO directories VALUES('/private/var/mobile/Containers/Data/Application/928F499E-48C7-43BB-B3E6-8D585E6FA011/Documents/');
```

Haven't dug into this more deeply, but perhaps there's some step in relinking that my implementation in this PR misses?

Reply:
What makes it more curious is that my branch does contain #12462, so I'm not sure why it's not updating the tracks

Reply:
Very strange. You should see this being logged
`Relocated tracks: [num of tracks]`
otherwise it fails silently earlier?

Reply:
Couldn't find that line in the log (after reinstalling), but found something else that looks suspicious:

```
warning [Main] Library - Skipping access check for missing or invalid directory QFileInfo(/private/var/mobile/Containers/Data/Application/FD476962-4D00-48DE-8F20-ABF7E49791DA/Documents/)
critical [Main] DEBUG ASSERT: "!oldDirectory.endsWith('/')" in function std::pair<DirectoryDAO::RelocateResult, QList<RelocatedTrack>> DirectoryDAO::relocateDirectory(const QString &, const QString &) const at src/library/dao/directorydao.cpp:220
```

Notably this is the old sandbox prefix. Maybe Mixxx doesn't like to relink something that no longer exists in the fs?

Reply:
Yeah, that's probably it! (the debug assert)
> Maybe Mixxx doesn't like to relink something that no longer exists in the fs?

well, that's the entire point of relocating, no ; )

(oh, you build with DEBUG_ASSERTIONS_FATAL=OFF )

Reply:
Does this log line already have the trailing / ?
`Relocating directory in internal track collection: oldDir -> newDir`

Reply:
Yes:

```
debug [Main] TrackCollectionManager - Relocating directory in internal track collection: "/private/var/mobile/Containers/Data/Application/3892CA05-7DC8-4DC1-B535-BB6CF05744A3/Documents/" -> "/private/var/mobile/Containers/Data/Application/C5C1DF42-1036-4328-B1AF-0B96889A02F0/Documents/"
```

For reference, here's a full log:

[mixxx-ios.log](https://github.com/user-attachments/files/16652277/mixxx-ios.log)

Reply:
So it looks like we already store the directory with a trailing slash. Patching the check to accept an existing trailing slash fixes the debug assert, but for some reason the track locations are still not updated properly in the db (there are probably other assumptions going on, haven't dug in that deeply):

```diff
diff --git a/src/library/dao/directorydao.cpp b/src/library/dao/directorydao.cpp
index c1022eb47e..e815498960 100644
--- a/src/library/dao/directorydao.cpp
+++ b/src/library/dao/directorydao.cpp
@@ -217,8 +217,10 @@ std::pair<DirectoryDAO::RelocateResult, QList<RelocatedTrac
k>> DirectoryDAO::rel
     // Appending '/' is required to disambiguate files from parent
     // directories, e.g. "a/b.mp3" and "a/b/c.mp3" where "a/b" would
     // match both instead of only files in the parent directory "a/b/".
-    DEBUG_ASSERT(!oldDirectory.endsWith('/'));
-    const QString oldDirectoryPrefix = oldDirectory + '/';
+    QString oldDirectoryPrefix = oldDirectory;
+    if (!oldDirectory.endsWith('/')) {
+        oldDirectoryPrefix += '/';
+    }
     query.prepare(QStringLiteral(
             "SELECT library.id,track_locations.id,track_locations.location "
             "FROM library INNER JOIN track_locations ON "
```

Reply:
I have updated the directory DAO to accept an existing trailing slash (28ed67bcf8) and now relinking seems to work properly without rescanning 🎉